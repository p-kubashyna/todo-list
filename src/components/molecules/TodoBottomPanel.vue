<script setup>
import TaskCounter from '@/components/atoms/TaskCounter.vue'
import FilterButton from '@/components/atoms/FilterButton.vue'
import ClearCompletedButton from '@/components/atoms/ClearCompletedButton.vue'
import { useStore } from 'vuex'

const store = useStore()

function setFilter (filterName) {
  store.commit('setFilter', filterName)
}

function deleteCompletedTasks () {
  store.commit('removeAllCompleted')
}
</script>

<template>
  <div class=bottom-panel>
    <TaskCounter/>
    <FilterButton @click="setFilter('incomplete')" label="Show incomplete" filter-name="incomplete"/>
    <FilterButton @click="setFilter('completed')" label="Show completed" filter-name="completed"/>
    <FilterButton @click="setFilter('all')" label="Show all" filter-name="all"/>
    <ClearCompletedButton @click="deleteCompletedTasks"/>
  </div>
</template>

<style>
.bottom-panel {
  display: flex;
  align-items: end;
  justify-content: space-evenly;
  height: 50px;
  background-color: white;
  padding-bottom: 10px;
}
</style>
