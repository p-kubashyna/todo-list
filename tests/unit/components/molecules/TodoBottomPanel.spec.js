import { describe, expect, it } from "@jest/globals";
import { mount } from "@vue/test-utils";
import TodoBottomPanel from "@/components/molecules/TodoBottomPanel.vue";
import { createStore } from "vuex";
import { FILTERS } from "@/constants/constants";

function findByText(wrap, text) {
  return wrap
    .findAll("button")
    .filter((n) => n.text().match(text))
    .at(0);
}

describe("Given the TodoBottomPanel component", () => {
  const defaultStore = {
    state: {
      activeFilterName: FILTERS.ALL,
    },
    getters: {
      getCompletedTodosCount: () => 0,
    },
  };

  it.each([
    [FILTERS.ALL, 1],
    [FILTERS.COMPLETED, 0],
    [FILTERS.INCOMPLETE, 0],
  ])("should render", (activeFilter, completedCount) => {
    const store = createStore({
      state: {
        activeFilterName: activeFilter,
      },
      getters: {
        getCompletedTodosCount: () => completedCount,
      },
    });
    const wrapper = mount(TodoBottomPanel, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it.each([
    ["Show incomplete", FILTERS.INCOMPLETE],
    ["Show completed", FILTERS.COMPLETED],
    ["Show all", FILTERS.ALL],
  ])("should change filter on click", async (buttonLabel, filterName) => {
    const store = createStore(defaultStore);
    store.commit = jest.fn();
    const wrapper = mount(TodoBottomPanel, {
      global: {
        plugins: [store],
      },
    });
    await findByText(wrapper, buttonLabel).trigger("click");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(1);
    expect(store.commit).toHaveBeenCalledWith("setFilter", filterName);
  });

  it("should delete completed on click", async () => {
    const store = createStore({
      state: {
        activeFilterName: FILTERS.ALL,
      },
      getters: {
        getCompletedTodosCount: () => 1,
      },
    });
    store.commit = jest.fn();
    const wrapper = mount(TodoBottomPanel, {
      global: {
        plugins: [store],
      },
    });
    await findByText(wrapper, "Delete completed").trigger("click");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(1);
    expect(store.commit).toHaveBeenCalledWith("removeAllCompleted");
  });

  it("should do nothing on click when delete completed button is disabled", async () => {
    const store = createStore(defaultStore);
    store.commit = jest.fn();
    const wrapper = mount(TodoBottomPanel, {
      global: {
        plugins: [store],
      },
    });
    await findByText(wrapper, "Delete completed").trigger("click");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(0);
  });
});
