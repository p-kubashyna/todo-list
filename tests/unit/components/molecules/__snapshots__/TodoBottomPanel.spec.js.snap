// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TodoBottomPanel component should render 1`] = `
<div
  class="bottom-panel"
>
  <span
    class="counter-label"
  >
    undefined tasks left
  </span>
  <button
    class="default-button active-button"
  >
    Show incomplete
  </button>
  <button
    class="default-button active-button"
  >
    Show completed
  </button>
  <button
    class="default-button active-filter active-button"
  >
    Show all
  </button>
  <button
    class="default-button active-button"
  >
    Delete completed
  </button>
</div>
`;

exports[`Given the TodoBottomPanel component should render 2`] = `
<div
  class="bottom-panel"
>
  <span
    class="counter-label"
  >
    undefined tasks left
  </span>
  <button
    class="default-button active-button"
  >
    Show incomplete
  </button>
  <button
    class="default-button active-filter active-button"
  >
    Show completed
  </button>
  <button
    class="default-button active-button"
  >
    Show all
  </button>
  <button
    class="default-button"
    disabled=""
  >
    Delete completed
  </button>
</div>
`;

exports[`Given the TodoBottomPanel component should render 3`] = `
<div
  class="bottom-panel"
>
  <span
    class="counter-label"
  >
    undefined tasks left
  </span>
  <button
    class="default-button active-filter active-button"
  >
    Show incomplete
  </button>
  <button
    class="default-button active-button"
  >
    Show completed
  </button>
  <button
    class="default-button active-button"
  >
    Show all
  </button>
  <button
    class="default-button"
    disabled=""
  >
    Delete completed
  </button>
</div>
`;
